File: articles/statistics-uncertainty-measurement-biology/r/variance_components.R
# Variance components workflow in R.

data_path <- file.path("data", "biological_technical_replicates.csv")
if (!file.exists(data_path)) {
  data_path <- file.path("..", "data", "biological_technical_replicates.csv")
}

data <- read.csv(data_path)

unit_means <- aggregate(measurement ~ biological_unit, data = data, FUN = mean)
unit_vars <- aggregate(measurement ~ biological_unit, data = data, FUN = var)

between_unit_variance <- var(unit_means$measurement)
within_unit_variance <- mean(unit_vars$measurement)

summary_df <- data.frame(
  between_biological_unit_variance = between_unit_variance,
  within_technical_variance = within_unit_variance,
  variance_ratio_biological_to_technical = between_unit_variance / within_unit_variance
)

print(round(summary_df, 5))

File: articles/statistics-uncertainty-measurement-biology/python/variance_components.py
"""
Variance components workflow.

Run:
    python python/variance_components.py
"""

from pathlib import Path

import pandas as pd


ARTICLE_DIR = Path(__file__).resolve().parents[1]
DATA_PATH = ARTICLE_DIR / "data" / "biological_technical_replicates.csv"


def main() -> None:
    data = pd.read_csv(DATA_PATH)

    unit_means = data.groupby("biological_unit")["measurement"].mean()

    between_unit_variance = unit_means.var(ddof=1)
    within_unit_variance = data.groupby("biological_unit")["measurement"].var(ddof=1).mean()

    summary = pd.DataFrame(
        {
            "between_biological_unit_variance": [between_unit_variance],
            "within_technical_variance": [within_unit_variance],
            "variance_ratio_biological_to_technical": [between_unit_variance / within_unit_variance],
        }
    )

    print(summary.round(5).to_string(index=False))


if __name__ == "__main__":
    main()
